age/keysource.go

```diff
@@ -0,0 +1,196 @@
+package age
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"filippo.io/age"
+)
+
+// MasterKey is an age key used to encrypt and decrypt sops' data key.
+type MasterKey struct {
+	Identity     string // a Bech32-encoded private key
+	Recipient    string // a Bech32-encoded public key
+	EncryptedKey string // a sops data key encrypted with age
+
+	parsedIdentity  *age.X25519Identity  // a parsed age private key
+	parsedRecipient *age.X25519Recipient // a parsed age public key
+}
+
+// Encrypt takes a sops data key, encrypts it with age and stores the result in the EncryptedKey field.
+func (key *MasterKey) Encrypt(datakey []byte) error {
+	buffer := &bytes.Buffer{}
+
+	if key.parsedRecipient == nil {
+		parsedRecipient, err := age.ParseX25519Recipient(key.Recipient)
+
+		if err != nil {
+			return fmt.Errorf("failed to parse input as Bech32-encoded age public key: %v", err)
+		}
+
+		key.parsedRecipient = parsedRecipient
+	}
+
+	w, err := age.Encrypt(buffer, key.parsedRecipient)
+
+	if err != nil {
+		return fmt.Errorf("failed to open file for encrypting sops data key with age: %v", err)
+	}
+
+	if _, err := w.Write(datakey); err != nil {
+		return fmt.Errorf("failed to encrypt sops data key with age: %v", err)
+	}
+
+	if err := w.Close(); err != nil {
+		return fmt.Errorf("failed to close file for encrypting sops data key with age: %v", err)
+	}
+
+	key.EncryptedKey = buffer.String()
+
+	return nil
+}
+
+// EncryptIfNeeded encrypts the provided sops' data key and encrypts it if it hasn't been encrypted yet.
+func (key *MasterKey) EncryptIfNeeded(datakey []byte) error {
+	if key.EncryptedKey == "" {
+		return key.Encrypt(datakey)
+	}
+
+	return nil
+}
+
+// EncryptedDataKey returns the encrypted data key this master key holds.
+func (key *MasterKey) EncryptedDataKey() []byte {
+	return []byte(key.EncryptedKey)
+}
+
+// SetEncryptedDataKey sets the encrypted data key for this master key.
+func (key *MasterKey) SetEncryptedDataKey(enc []byte) {
+	key.EncryptedKey = string(enc)
+}
+
+// Decrypt decrypts the EncryptedKey field with the age identity and returns the result.
+func (key *MasterKey) Decrypt() ([]byte, error) {
+	ageKeyDir, ok := os.LookupEnv("SOPS_AGE_KEY_DIR")
+
+	if !ok {
+		userConfigDir, err := os.UserConfigDir()
+
+		if err != nil {
+			return nil, fmt.Errorf("user config directory could not be determined: %v", err)
+		}
+
+		ageKeyDir = filepath.Join(userConfigDir, ".sops", "age")
```

Comment:
Directories in userConfigDir can usually omit the `.`. For example $HOME/.config/sops instead of $HOME/.config/.sops.
```suggestion
		ageKeyDir = filepath.Join(userConfigDir, "sops", "age")
```

Reply:
Actually I just found out that age looks up keys in `$HOME/.age/keys.txt`, `~/.ssh/id_rsa` and `~/.ssh/id_ed25519`.

https://github.com/FiloSottile/age/blob/189041b668629795593766bcb8d3f70ee248b842/cmd/age/age.go#L243

I think sops should just do the same instead of inventing its own locations. The environment variable is fine though.

Reply:
Also it seems to collect just all keys from those files instead of using the the public key id as a name selector.

Reply:
Also this allows to have the private key encrypted:

https://github.com/FiloSottile/age/blob/189041b668629795593766bcb8d3f70ee248b842/cmd/age/parse.go#L97